src/devices/Hmc5883l/Hmc5883l.cs

```diff
@@ -0,0 +1,123 @@
+        private readonly byte _outputRate;
+        private readonly byte _gain;
+
+        /// <summary>
+        /// HMC5883L Raw Vector
+        /// </summary>
+        public Vector3 RawVector => ReadRawVector();
+
+        /// <summary>
+        /// HMC5883L Heading (DEG)
+        /// </summary>
+        public double Heading => VectorToAngle(ReadRawVector());
+
+        /// <summary>
+        /// Initialize a new HMC5883L device connected through I2C
+        /// </summary>
+        /// <param name="sensor">I2C Device, like UnixI2cDevice or Windows10I2cDevice</param>
+        /// <param name="gain">Gain Setting</param>
+        /// <param name="measuringMode">The Mode of Measuring</param>
+        /// <param name="outputRate">Typical Data Output Rate (Hz)</param>
+        public Hmc5883l(I2cDevice sensor, Gain gain = Gain.Gain1090, MeasuringMode measuringMode = MeasuringMode.Continuous, OutputRate outputRate = OutputRate.Rate15)
+        {
+            _sensor = sensor;
+            _gain = (byte)gain;
+            _measuringMode = (byte)measuringMode;
+            _outputRate = (byte)outputRate;
+
+            Initialize();
+        }
+
+        /// <summary>
+        /// Initialize the sensor
+        /// </summary>
+        private void Initialize()
+        {
+            // Details in Datasheet P12
+            byte configA = (byte)(0x70 + _outputRate << 2);
+            byte configB = (byte)(_gain << 5);
+
+            _sensor.Write(new[] { (byte)Register.HMC_CONFIG_REG_A_ADDR, configA });
+            _sensor.Write(new[] { (byte)Register.HMC_CONFIG_REG_B_ADDR, configB });
+            _sensor.Write(new[] { (byte)Register.HMC_MODE_REG_ADDR, _measuringMode });
+        }
+
+        /// <summary>
+        /// Read raw data from HMC5883L
+        /// </summary>
+        /// <returns>Raw Data</returns>
+        private Vector3 ReadRawVector()
+        {
+            Span<byte> xRead = stackalloc byte[2];
+            Span<byte> yRead = stackalloc byte[2];
+            Span<byte> zRead = stackalloc byte[2];
+
+            _sensor.WriteByte((byte)Register.HMC_X_MSB_REG_ADDR);
+            _sensor.Read(xRead);
+            _sensor.WriteByte((byte)Register.HMC_Y_MSB_REG_ADDR);
+            _sensor.Read(yRead);
+            _sensor.WriteByte((byte)Register.HMC_Z_MSB_REG_ADDR);
+            _sensor.Read(zRead);
+
+            short x = BinaryPrimitives.ReadInt16BigEndian(xRead);
+            short y = BinaryPrimitives.ReadInt16BigEndian(yRead);
+            short z = BinaryPrimitives.ReadInt16BigEndian(zRead);
+
+            return new Vector3(x, y, z);
+        }
+
+        /// <summary>
+        /// Calculate angle
+        /// </summary>
+        /// <param name="vector">HMC5883L Raw Data</param>
+        /// <param name="isRad">Determine whether the angle of return is RAD or DEG. Default value is DEG.</param>
+        /// <returns>Angle</returns>
+        public double VectorToAngle(Vector3 vector, bool isRad = false)
```

Comment:
private

Reply:
also why are you adding PI to the result?

Reply:
@krwq The range of heading is 0-360 deg. If 180 deg are not added, the range will be -180-180 deg. 

Now that this method is private, there is no need to consider too many details, as long as the calculation of the heading is correct. Other calculations are handed over to the person using the sensor.

Reply:
@ZhangGaoxing in that case you should add 360 and modulo 360 - otherwise you change the direction. What's the problem with negative numbers here? For radians it is usual that negative numbers are there

Reply:
(unfortunately the modulo which someone has invented does not match the math and works incorrectly for negative numbers, that's why you add 360 first)

Reply:
@krwq For instance, if I get an angle -180 deg, it means I'm facing north, but the North Angle is 0 deg, so I need to add 180 deg. I don't think it's a math problem. I'm just correcting the results.

To avoid ambiguity, how about changing the method name to VectorToHeading and remove rad conversion? Because this is a private method, the purpose of writing this method is only to compute heading.😄

Reply:
Other calculations are handed over to the person using the sensor.